js/modules/loki_snode_api.js

```diff
@@ -0,0 +1,170 @@
+/* global log, window, Whisper */
+
+const fetch = require('node-fetch');
+const is = require('@sindresorhus/is');
+const dns = require('dns');
+
+// Will be raised (to 3?) when we get more nodes
+const MINIMUM_SWARM_NODES = 1;
+
+class LokiSnodeAPI {
+
+  constructor({ url, swarmServerPort }) {
+    if (!is.string(url)) {
+      throw new Error('WebAPI.initialize: Invalid server url');
+    }
+    this.url = url;
+    this.swarmServerPort = swarmServerPort
+      ? `:${swarmServerPort}`
+      : '';
+    this.swarmsPendingReplenish = {};
+    this.ourSwarmNodes = {};
+  }
+
+  getRandomSnodeAddress() {
+    /* resolve random snode */
+    return new Promise((resolve, reject) => {
+      dns.resolveCname(this.url, (err, address) => {
+        if(err) {
+          reject(err);
+        } else {
+          resolve(address[0]);
+        }
+      });
+    });
+  }
+
+  unreachableNode(pubKey, nodeUrl) {
+    if (pubKey === window.textsecure.storage.user.getNumber()) {
+      delete this.ourSwarmNodes[nodeUrl];
+    }
+  }
+
+  updateLastHash(nodeUrl, hash) {
+    if (!this.ourSwarmNodes[nodeUrl]) {
+      this.ourSwarmNodes[nodeUrl] = {
+        lastHash: hash,
+      }
+    } else {
+      this.ourSwarmNodes[nodeUrl].lastHash = hash;
+    }
+  }
+
+  async getOurSwarmNodes() {
+    if (
+      !this.ourSwarmNodes ||
+      Object.keys(this.ourSwarmNodes).length < MINIMUM_SWARM_NODES
+    ) {
+      // Try refresh our swarm list once
+      const ourKey = window.textsecure.storage.user.getNumber();
+      const nodeAddresses = await window.LokiSnodeAPI.getSwarmNodes(ourKey);
+
+      this.ourSwarmNodes = {};
+      nodeAddresses.forEach(url => {
+        this.ourSwarmNodes[url] = {};
+      })
+      if (!this.ourSwarmNodes || Object.keys(this.ourSwarmNodes).length === 0) {
```

Comment:
nit: move this check a couple of lines up:

```js
const nodeAddresses = await window.LokiSnodeAPI.getSwarmNodes(ourKey);
if (nodesAddresses.length === 0) {
  throw ...
}

this.ourSwarmNodes = {};
...
```
(Probably best checking that `nodeAddresses` is always an array tho)